### ISSUE TITLE
Cross-compilation issue with linker search paths

### ISSUE THREAD

**Issue by kognise** [NixOS/nixpkgs#139966]
### Describe the bug
I am encountering an issue when cross-compiling with a Windows library installed, `libpthread` in this example. GCC is incorrectly picking up the Windows library instead of the system one.

I came across this in the context of a Rust project with a `build.rs` file. First, the build file has to be compiled for and executed on the host platform with GCC. Afterward, the rest of the project can be compiled for the target with MinGW.

`NIX_LDFLAGS` contains the Windows library when it's installed, which causes GCC to error because it doesn't obviously understand the format. If I manually remove it from the environment, then MinGW doesn't know where to pick up the library and the build fails.

### Steps To Reproduce
1. Clone https://github.com/kognise/crosscomp-mcve
2. Run `nix-shell --pure`
3. Run `cargo clean && cargo build --target x86_64-pc-windows-gnu`

### Expected behavior
I'm relatively new to Nix, so I honestly don't know what the right solution to this is.

The expected behavior *from a developer's perspective* is simply for GCC to find the Linux `libpthread` and for MinGW to find the Windows version so that everything can compile nicely. 

### Screenshots
![](https://doggo.ninja/KeWZrV.png)

![](https://doggo.ninja/mlXA3L.png)

### Additional context
I would just like to mention that this isn't a problem with Rust exclusively, or the overlay I'm using. This is just the simplest way I've been able to reproduce it, and how I came across the issue during real-world development.

### Notify maintainers
Not sure who I should specifically ping for this, so apologies if I incorrectly mentioned you.

@matthewbauer (from MinGW derivation history)

@madjar @cstrahan @globin @havvy (from rustc maintainers, probably didn't need to ping all of you so sorry)

### Metadata
- system: `"x86_64-linux"`
- host os: `Linux 5.10.67, NixOS, 21.05.3367.fd8a7fd07da (Okapi)`
- multi-user?: `yes`
- sandbox: `yes`
- version: `nix-env (Nix) 2.3.15`
- channels(root): `"nixos-21.05.3367.fd8a7fd07da"`
- channels(kognise): `"nixos-21.05-21.05.3367.fd8a7fd07da"`
- nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by lammermann** [comment]
I had a similar problem in one of my projects. The advice in this link helped me https://github.com/nix-community/naersk/issues/181#issuecomment-874352470 . Hope it works for you too.

**Comment by chloekek** [comment]
A way to work around this is to not put `windows.pthreads` in the build inputs of the shell, but assign it to an environment variable in the shell, e.g. `WINDOWS_PTHREADS = windows.pthreads;`. Then, in `build.rs`, pass that environment variable to rustc with `println!("cargo:rustc-link-search=native={}/lib", env::var("WINDOWS_PTHREADS").unwrap());`.

**Comment by RossSmyth** [comment]
This is still a really annoying issue :(

The way I solved it is by doing
`RUSTFLAGS = libs.optional stdenv.hostPlatform.isMinGW "-L native=${windows.pthreads}/lib";`

This line
https://github.com/NixOS/nixpkgs/blob/8bb37161a0488b89830168b81c48aed11569cb93/pkgs/build-support/rust/build-rust-package/default.nix#L171
Doesn't work. I do not know why.

**Comment by pinage404** [comment]
i have been able to build Rust targeting Windows with this

```nix
{
  contents = [
    rust-toolchain
    pkgs.stdenv.cc
    (pkgs.pkgsCross.mingwW64.stdenv.cc.override ({
      extraBuildCommands = ''
        printf '%s' '-L ${pkgs.pkgsCross.mingwW64.windows.mingw_w64_pthreads}/lib' >>$out/nix-support/cc-ldflags
      '';
    }))
  ];
}
```

https://gitlab.com/pinage404/git-gamble/-/blob/ac671d9fa2bd880e2adb596718facdb5f0519cb3/nix/packages/container-image-rust-windows/default.nix#L27

**Comment by RossSmyth** [comment]
Yeah that should be equivalent, but really one shouldn't have to do these workarounds. They are pervasive if you look around nixpkgs and github just to get around this issue.

**Comment by Sewer56** [comment]
It would be ideal if such workaround could be avoided- that is true, unfortunately.

I guess it's just another case of

> I have to write a flake from scratch- before I can build or run random <x> project from the web, and since the repo probably wouldn't accept a flake.nix for obscure OS in its repository, I have to do it for every project I touch from scratch every time.

Or however you want to phrase it.
